### DISCUSSION TITLE
Discrepancy between high precision,recall and low average IOU (YOLO11, Detect)

### DISCUSSION THREAD

**Discussion by JohnLYYN123** [ultralytics/ultralytics#22387] [Q&A] [answered]
Hello, 

I’m running validation on a custom dataset using YOLO11, and I’ve noticed an unexpected pattern related to the conf and save_json parameters.

When I validate with different confidence thresholds (conf values ranging from 0.001 to 0.5), I consistently get high precision (~0.9) and high recall (~0.85), regardless of the confidence setting.

However, when I compute the average IoU from the bounding boxes in the generated predictions.json file, I find that the average IoU values are relatively low, especially for small confidence thresholds (e.g., average IoU ≈ 0.35 at conf=0.001). 

This raises a question:

How can YOLO report such high precision and recall while the average IoU among the saved predictions in predictions.json remains low?

To clarify:
	•	The IoU values are computed directly from the coordinates in the exported JSON.
	•	Validation is run using the official Ultralytics model.val() API with single-class evaluation.
	•	I’ve confirmed that the JSON export includes all predictions filtered by the specified conf threshold.
	• I only have one class.

I’d appreciate any insight into how precision/recall are computed internally during validation, and whether they use a different IoU or confidence filtering scheme than the JSON export.

Thank you very much!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @JohnLYYN123, thanks for opening this discussion and for using YOLO11 Detect with a custom dataset 🚀. This is an automated response to help you get started while an Ultralytics engineer reviews your post and will assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find detailed [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, along with explanations of validation metrics and exports that may clarify the behavior you're observing.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including:
-Exact validation command or Python snippet used (including `conf`, `save_json`, `iou`, `single_cls`, and any `nms`-related args)  
-Dataset details (single-class confirmation and any class mapping)  
-Sample of your `predictions.json` and a few image-level examples illustrating the discrepancy  
-Full validation logs and environment info  
And verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Precision and recall are calculated using a fixed IoU of 0.5 and then finding how many TPs, FPs and FNs are obtained at that IoU. 

The average IoU at confidence 0.001 will be low because it involves a lot of incorrect detections.

You should check the average IoU at higher confidence levels.

**Reply by JohnLYYN123** [reply]
Hi

Thank you so much for the reply!

I understand that higher confidence thresholds can lead to better average IoUs. However, what really confuses me is that the precision and recall remain consistently high regardless of how low I set the confidence threshold during validation.

For example, when I validated with conf=0.001 and conf=0.3, I obtained around 0.9 precision and 0.85 recall in both cases. However, the average IoU computed from the predictions.json file differed significantly, 0.384 for conf=0.001 vs 0.721 for conf=0.3.

Since IoU determines the number of TP, FP, and FN, a low average IoU (e.g., 0.384) should logically correspond to more FP and FN, leading to lower precision and recall. So I’m confused as to why the reported precision and recall from the validation mode stay high even when the average IoU is low.

In [Ultralytics' documentation](https://docs.ultralytics.com/modes/val/#arguments-for-yolo-model-validation), conf is described to affect the number of FPs and recall if conf is set low. However, this doesn’t seem to align with the results I’m seeing across different conf values.

I have two related questions:
	1.	Are the reported precision and recall values computed based on the same set of predictions that are exported to the predictions.json file?
	2.	Does the validation process use any internal mechanism (such as sweeping through multiple confidence thresholds or internally filtering low-confidence detections) when computing precision and recall, instead of strictly using the user-specified conf value?

Thanks!

**Reply by Y-T-G** [reply]
You can read this

https://github.com/ultralytics/ultralytics/issues/17047#issuecomment-2424660435

**Reply by JohnLYYN123** [reply]
Thank you very much again!

So, may I simply assume that the conf value we provide mainly controls the number of predictions included in the evaluation process and in the predictions.json file?

Also, since the precision and recall values shown in the log and in the confusion matrix are calculated differently, and the ones in the log are based on the confidence threshold that yields the best F1 score, which set of precision and recall values would you recommend using for reporting results in my project?

Finally, where can I find the exact confidence threshold that corresponds to the best F1 score? Does this threshold correspond to the one shown in the BoxF1_curve.png file generated during validation?

Thanks again for your time and clarification!

**Reply by glenn-jocher** [reply, marked as answer]
Great questions—short answers below.

1) Yes: during `val`, `conf` acts as a lower bound to collect detections for building the PR/F1 curves and for `predictions.json`; the reported P/R are computed by sweeping confidence thresholds and selecting the F1‑optimal point, so they can remain stable across different `conf` minima. See the F1/PR curves explanation in the [Performance metrics deep dive](https://docs.ultralytics.com/guides/yolo-performance-metrics/), and note that JSON export simply serializes the detections provided to it ([JSON export for COCO evaluation](https://docs.ultralytics.com/reference/models/yolo/detect/val/#pred_to_json)).

2) For reporting, we recommend mAP50‑95 as the primary metric, and if you include P/R, report the values at the F1‑optimal threshold and state that threshold; the confusion matrix may differ because it uses its own filtering (by default it applies `conf=0.25` when the validation `conf` is left at its default), see the filtering logic in [process_batch confusion‑matrix filtering](https://docs.ultralytics.com/reference/utils/metrics/#process_batch). Matching of predictions to labels is done across IoU thresholds internally, as shown in [IoU matching across thresholds](https://docs.ultralytics.com/reference/engine/validator/#match_predictions).

3) You can retrieve the exact F1‑optimal confidence from the returned metrics; it corresponds to the peak in the generated `F1_curve.png`. Example:

```python
from ultralytics import YOLO
import numpy as np

model = YOLO("yolo11n.pt")
metrics = model.val(data="your.yaml", save_json=True)

i = int(metrics.box.f1_curve.argmax())
best_conf = float(metrics.box.px[i])
print(f"Best F1 confidence: {best_conf:.3f}")

# To align predictions.json with reported P/R:
model.val(data="your.yaml", conf=best_conf, save_json=True)
```

This should align your IoU analysis with the P/R reported by validation.

**Reply by JohnLYYN123** [reply]
Hello,

Thank you so much for the detailed explanation. 

You guys are amazing and I really appreciate the help! Wish you all the best!

**Reply by glenn-jocher** [reply]
Thanks John—glad it helped; credit goes to the YOLO community and Ultralytics team. If this resolves your question please mark the discussion as answered; for clean reporting you can persist validation outputs from `model.val()` using `metrics.to_df()` or `metrics.to_json()` for your records.

**Reply by JohnLYYN123** [reply]
Hello, 

Thank you so much!

I have a continued question from the previous answers. It is still related to the best confidence. After using the code you provided, I was able to obtain the best confidence but this confidence is different from the reported best confidence appear in the BoxF1_curve.png. For instance, the best confidence I got from the code was 0.294 and the reported confidence in the BoxF1_curve.png was 0.317. I only have one class. 

Could you please let me know why this difference occurs, and which value should be considered the correct one for evaluation?

Thanks in advance for your help!

**Reply by glenn-jocher** [reply]
Great follow-up—this gap is expected. The value printed on BoxF1_curve.png is chosen after smoothing the class-mean F1 curve (smoothing=0.1) and then taking the argmax, while your snippet used the raw curve. To match the figure exactly, compute the index on the smoothed mean curve and read the confidence from `px`; use this smoothed value for reporting since it drives the logged P/R/F1.

```python
from ultralytics import YOLO
from ultralytics.utils.metrics import smooth

model = YOLO("path/to/best.pt")
m = model.val(data="your.yaml", plots=True)

i = int(smooth(m.box.f1_curve.mean(0), 0.1).argmax())
best_conf = float(m.box.px[i])
print(f"Best F1 confidence (matches figure): {best_conf:.3f}")
```

This mirrors the library’s internal selection in `ap_per_class` (see the returned `f1_curve` and the smoothed `argmax` logic in the [metrics reference](https://docs.ultralytics.com/reference/utils/metrics/#ap_per_class)).

**Reply by JohnLYYN123** [reply]
Thank you so much for the detailed reply!

**Reply by glenn-jocher** [reply]
You're welcome—if this resolves things, please mark the discussion as answered; note the F1‑optimal confidence is dataset/split‑specific, so recompute it on your final val/test set and use that value for both reporting and for `predict()`/`val` (`conf=<best_conf>`) to keep JSON, curves, and logs consistent.
